<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate Marker Genes Proxy — FindMarkers.proxy • DecoFlex</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate Marker Genes Proxy — FindMarkers.proxy"><meta property="og:description" content="The FindMarkers.proxy function is a flexible interface for identifying
marker genes using various methods.It's designed to be readily extensible,
facilitating the integration of novel marker calculation techniques in the
future. The function calculates the markers for a specific group relative to
all other cells (rest_groups) in a given Seurat object (seurat.reference)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DecoFlex</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">References</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    functions
  </a>
</li>
<li>
  <a href="https://github.com/crhisto/DecoFlex" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate Marker Genes Proxy</h1>
    <small class="dont-index">Source: <a href="https://github.com/crhisto/DecoFlex/blob/HEAD/R/tree_guided_deconvolution.R" class="external-link"><code>R/tree_guided_deconvolution.R</code></a></small>
    <div class="hidden name"><code>FindMarkers.proxy.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>FindMarkers.proxy</code> function is a flexible interface for identifying
marker genes using various methods.It's designed to be readily extensible,
facilitating the integration of novel marker calculation techniques in the
future. The function calculates the markers for a specific group relative to
all other cells (rest_groups) in a given Seurat object (seurat.reference).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S3 method for proxy</span></span>
<span><span class="fu">FindMarkers</span><span class="op">(</span></span>
<span>  <span class="va">seurat.reference</span>,</span>
<span>  <span class="va">group</span>,</span>
<span>  <span class="va">rest_groups</span>,</span>
<span>  test.use.value <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>  marker_strategy <span class="op">=</span> <span class="st">"keep_default_values"</span>,</span>
<span>  max.size.percentage <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  param.logfc.threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>seurat.reference</dt>
<dd><p>A Seurat object used as the reference for marker gene
calculation. This object contains the single-cell RNA-seq data.</p></dd>


<dt>group</dt>
<dd><p>A string specifying the cell group (based on clustering) for
which to calculate marker genes.</p></dd>


<dt>rest_groups</dt>
<dd><p>A list or vector specifying the remaining cell groups to
be considered as background in the marker genes calculation.</p></dd>


<dt>test.use.value</dt>
<dd><p>String specifying the statistical test to use for
marker identification. The methods currently supported include:
'wilcox', 'bimod', 't', 'negbinom', 'poisson', 'LR', and 'MAST'. Please note
that 'roc' and 'DESeq2' methods are currently not supported due to
discrepancies in output values and support issues, respectively. Default is
'wilcox'.</p></dd>


<dt>marker_strategy</dt>
<dd><p>A string specifying the strategy to use in marker
gene calculation.
'keep_default_values' (default) retains the default values. Other strategies
may be added in the future.</p></dd>


<dt>max.size.percentage</dt>
<dd><p>Maximum acceptable size (as a proportion) for a
group in the marker gene calculation. Default is 0.70.</p></dd>


<dt>param.logfc.threshold</dt>
<dd><p>Threshold for the log-fold change, which
determines the minimum change required for a gene to be considered as a
marker. Default is 2.0.</p></dd>


<dt>verbose</dt>
<dd><p>Logical value indicating whether to display detailed progress
messages during function execution. Default is FALSE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list of marker genes for the specified cell group.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function serves as a proxy to different methods for marker gene
calculation. It currently supports several statistical tests and provides a
convenient point of extension for the inclusion of additional marker
calculation methods. Note that the support for different tests can be
influenced by the characteristics of the single-cell experiments, and it's
important to choose a method that's appropriate for your data.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Crhistian Cardona.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

